### ISSUE TITLE
Crash if an object is destroyed during a multiple collision check

### ISSUE THREAD

**Issue by Alex-Canales** [HaxeFlixel/flixel#2045]
- **Haxe version:** 3.2.1
- **Flixel version:** 4.2.1
- **OpenFL version:** 3.6.1
- **Lime version:** 2.9.1
- **Affected targets:** All

When a sprite ``A`` collides multiple sprites: if ``A`` is destroyed after the first collision, it will be null for the other collision, a ``Null Object Reference`` will appear and the program will crash.

The reproducible bug:

```haxe
package;

import flixel.FlxSprite;

class PlayState extends flixel.FlxState
{
    private var bullet:FlxSprite;

    override public function create():Void
    {
        super.create();

        // Context: the player shoots to the enemy but the bullet overlaps the
        // player and the enemy.

        // The order is important: if in another order it may not crash.
        var enemy:FlxSprite = new FlxSprite(150, 70);
        enemy.makeGraphic(10, 12, flixel.util.FlxColor.BLUE);
        enemy.setSize(10, 12);
        add(enemy);

        bullet = new FlxSprite(155, 81);
        bullet.makeGraphic(4, 4, flixel.util.FlxColor.ORANGE);
        bullet.setSize(4, 4);
        bullet.velocity.set(0, -1);
        // With the following velocity, there is no collision so no crash.
        // bullet.velocity.set(-1, 0);
        // bullet.velocity.set(1, 0);
        // bullet.velocity.set(0, 1);
        add(bullet);

        var player:FlxSprite = new FlxSprite(150, 83);
        player.makeGraphic(10, 12, flixel.util.FlxColor.GREEN);
        player.setSize(10, 12);
        add(player);
    }

    override public function update(elapsed:Float):Void
    {
        super.update(elapsed);
        flixel.FlxG.collide(bullet, this, collide);
    }

    private function collide(bullet:FlxSprite, other:FlxSprite):Void
    {
        bullet.destroy();
    }

}
```

The log when compiled in C++ for Linux (using ``haxelib run lime test linux -debug``)

```
Null Object Reference
Called from openfl._legacy.display.Stage.__pollTimers (openfl/_legacy/display/Stage.hx line 1084)
Called from openfl._legacy.display.Stage.__checkRender (openfl/_legacy/display/Stage.hx line 351)
Called from openfl._legacy.display.Stage.__render (openfl/_legacy/display/Stage.hx line 1103)
Called from openfl._legacy.display.DisplayObjectContainer.__broadcast (openfl/_legacy/display/DisplayObjectContainer.hx line 286)
Called from openfl._legacy.display.DisplayObject.__broadcast (openfl/_legacy/display/DisplayObject.hx line 161)
Called from openfl._legacy.display.DisplayObject.__dispatchEvent (openfl/_legacy/display/DisplayObject.hx line 182)
Called from openfl._legacy.events.EventDispatcher.dispatchEvent (openfl/_legacy/events/EventDispatcher.hx line 98)
Called from flixel.FlxGame.onEnterFrame (flixel/FlxGame.hx line 554)
Called from flixel.FlxGame.step (flixel/FlxGame.hx line 689)
Called from flixel.FlxGame.update (flixel/FlxGame.hx line 765)
Called from flixel.FlxState.tryUpdate (flixel/FlxState.hx line 171)
Called from MenuState.update (MenuState.hx line 39)
Called from flixel.FlxG.overlap (flixel/FlxG.hx line 376)
Called from flixel.system.FlxQuadTree.execute (flixel/system/FlxQuadTree.hx line 593)
Called from flixel.system.FlxQuadTree.execute (flixel/system/FlxQuadTree.hx line 597)
Called from flixel.system.FlxQuadTree.execute (flixel/system/FlxQuadTree.hx line 605)
Called from flixel.system.FlxQuadTree.execute (flixel/system/FlxQuadTree.hx line 584)
Called from flixel.system.FlxQuadTree.overlapNode (flixel/system/FlxQuadTree.hx line 655)
Called from flixel.FlxObject.separate (flixel/FlxObject.hx line 83)
Called from flixel.FlxObject.separateX (flixel/FlxObject.hx line 190)
Called from flixel.FlxObject.computeOverlapX (flixel/FlxObject.hx line 113)
```

The JavaScript error when compiled in HTML5 (using ``haxelib run lime test html5 -debug``)

```
TypeError: Object1.last is null
http://localhost:3000/TestFlxG.js
Line 7153
```

If you have any questions, I would be happy to answer them as much as I can.

**Comment by Gama11** [comment]
This doesn't seem unexpected to me. You can't rely on a destroyed sprite still being usable.

Usually, you'd use `kill()` in collision callbacks rather than `destroy()` - the latter is for memory cleanup (during state switches etc).

**Comment by Alex-Canales** [comment]
Indeed, before submitting the bug I saw that it does not occur when ``kill`` is used instead of ``destroy``.

This "bug" maybe depends on the developer point of view:

- From a Haxe Flixel developer point of view, I understand this is a pretty defined behaviour: you should not try to do something with a destroyed object so this is normal if the program crashes.
- From an user point of view, this is "brutal": you check a collision, decide to remove the object since you do not need it after the collision, the program crashes no questions asked. This is not an easy bug to find since it is deeply connected with the hidden framework behaviour and does not happen, in the code, explicitly where you do the ``destroy``.

Maybe simply checking if the object is not null before using it and doing nothing if this is the case would be a "nicer" way to handle this problem. But then you are adding yet another operation in a part where performance is crucial.

**Comment by Gama11** [comment]
>But then you are adding yet another operation in a part where performance is crucial.

Not only that, there's countless places where checks like this would have to be added - basically anywhere  something is done with a `FlxObject`, it will fail if it has been destroyed.

**Comment by Alex-Canales** [comment]
OK, it will be really unproductive to check that everywhere. Maybe you should change the documentation on the ``destroy`` function to explicitly set that using it loosely can crash internally the framework. Reading its documentation, I would have never thought it can cause internal problems in the Haxe Flixel system when called in a callback function.

A question I have, not completely unrelated to the problem I faced. I two objects overlapping when created:

- There is no collision no object moves toward the other.
- There is no collision if one object is completely inside one.

Is it the expected behaviour? I was also surprised at first but I can understand why it would be like that. I prefer to ask here rather than opening a "wrong" issue.

**Comment by Gama11** [comment]
Agreed, the docs could maybe be a little better... There is a big "WARNING" already, but it doesn't really make clear what "remove this object entirely" means.. (plus "remove" itself is a bit misleading).

About the collision issue, it's hard to tell without seeing a concrete example.

**Comment by Alex-Canales** [comment]
I read again my last message: no wonder why you could not understand what I was saying. The more I speak English, the worse I speak it! I have edited it to make more sense.

About the collision issue, I slept on it before writing an issue about it. After thinking about it, the current collision behaviour makes perfect sense: there is no way to know how objects should be separated if one is already inside the other one before checking the collision.

In a nutshell: I have no issue about collision.

**Comment by Gama11** [comment]
The docs should be a bit better now.

**Comment by Alex-Canales** [comment]
Yes, it is more explicit now. Thank you for the documentation update.
